GSoC_19/posts/websocket-passive-scan-using-scripts-with-zap.org

```diff
@@ -0,0 +1,845 @@
+#+BEGIN_COMMENT
+.. title: WebSocket Passive Scan using scripts with ZAP
+.. slug: websocket-passive-scan-using-scripts-with-zap
+.. date: 2019-07-11 14:43:55 UTC+03:00
+.. tags: ZAP, tutorial, websockets, passive, scan
+.. category: 
+.. link: 
+.. description: 
+.. type: text
+
+#+END_COMMENT
+#+OPTIONS: toc:3 
+
+* Introduction 
+
+  In this blog post we are going to demonstrate a brand new feature of OWASP
+  Zed Attack Proxy (ZAP). Specifically, we are going to use WebSocket extension in order to
+  perform Passive Scanning at WebSocket connections. Even if you haven't been
+  familiar with ZAP, this guide will try to be extensively enough to cover the
+  basics behind ZAP and give you the ability perform and write WebSocket passive
+  scan rules. On other hand, if you have been already familiar with ZAP, you can
+  skip the introduction section.  
+
+  #+BEGIN_QUOTE
+  At its core, ZAP is what is known as "man-in-the-middle proxy". It stands
+  between the tester's browser and web application so that it can intercept and
+  inspect messages sent between browser and the web application.
+
+  from [[https://github.com/zaproxy/zaproxy/releases/download/v2.8.0/ZAPGettingStartedGuide-2.8.pdf][Getting Started Guide (pdf)]].
+  #+END_QUOTE
+
+  #+ATTR_HTML: :align center :width 400 height 200
+  [[../../images/pscan-tutorial/zap_as_proxy.png]] 
+
+** WebSocket Extension
+
+   By default WebSocket extension is installed and it is responsible to intercept
+   and inspect websocket messages and connections. Other functions like setting
+   breakpoints, modify the content if needed, and then forward those messages to
+   the destination are also available, however it is out of the scope for this
+   blog post. The WebSocket messages are displayed in a separate tab at the
+   bottom of the window. 
+ 
+   #+ATTR_HTML: :align center :width 600 height 600
+   #+CAPTION: WebSocket Panel
+   [[../../images/pscan-tutorial/websocket_panel_1.png]] 
+  
+** WebSocket Passive Scanning                   
+
+    #+BEGIN_QUOTE
+    The passive scans rules are used to warn the user of potential vulnerability
+    that can be detected passively - _they not allowed to send any new WebSocket_
+    _message or manipulate the messages in any way_. They typically run against
+    all of the WebSocket messages that flow thought ZAP. Passive rules run in
+    separate background thread so that they have as little effect on performance
+    as possible. 
+
+    /from [[https://zaproxy.blogspot.com/2014/04/hacking-zap-3-passive-scan-rules.html][ZAP blog]], (modified in order to fit at the WebSocket Passive scan case)/
+    #+END_QUOTE
+    
+    At the end, what we want from the WebSocket Passive scan is to check all
+    WebSocket message's payloads for information which might reveals the
+    existence of a vulnerability (e.x. an error message).
```

Comment:
- `of an issue/vulnerability`, not all alerts are (potential) vulnerabilities.
 - `an internal error`